### ISSUE TITLE
New plan for trace storage work

### ISSUE THREAD

**Issue by StachuDotNet** [darklang/stable-dark#10]
### DB Clone
- [ ] prototype the DB clone (go through the steps, record "down times")
  - [ ] `events` table (not v2)
    - [ ] verify that nothing is querying `events` (not v2) table
    - [ ] look at the `events` table and see if it's using any foreign keys
      - [ ] if any, remove the foreign keys
    - [ ] delete the events table  from `dark-west`

  - [ ] do the DB clone
  - [ ] update the DB clone
    - [ ] set zoning to single-zone (applies to both servers and storage)
    - [ ] update postgres version to 14(?)
    - [ ] turn down the CPUs by ~1/3

### Cloud storage
- [ ] delete trace-related tests
- [ ] check that 404s continue to work
- [ ] ensure we overwrite cloud storage traces for `execute_handler` button
- [ ] check if execute_function traces are appropriately merged with a cloud-storage -based trace
- [ ] garbage collection - set object lifecycle for bucket or for traces
- [ ] ensure pusher is supported
- [ ] do walkthrough and check it all works

### monitoring
- [ ] schedule weekly call/meeting where we review usage, for 4 weeks. at the end of such, consider what to do
  - [ ] check table sizes
  - [ ] check costs

### migrate existing canvases
- [ ] upload to both simultaneously
- [ ] fetch and upload existing trace data for existing canvases/handlers
- [ ] possibly automatically switch LD flag once this is done
- [ ] switch all users to only use uploaded storage data

Maybe later?
- [ ] turn on private IPs (requires DB downtime)

**Comment by StachuDotNet** [comment]
- toplevel_oplists has PK
- stored_events_v2 no PK
- function_results_v3 no PK
- function_arguments no PK
- traces_v0 has PK
- user_data has PK
- system_migrations has PK
- static_asset_deploys has PK
- secrets has PK
- scheduling_rules has PK
- packages_v0 has PK
- op_ctrs no PK (tho does have unique constraint)
- events_v0 has PK
- events has PK
- custom_domains has PK
- cron_records has PK
- canvases has PK
- accounts has PK
- access no PK

from an earlier call - notes on which tables have/don't have PKs

**Comment by StachuDotNet** [comment]
Coordinating downtime to reduce costs

## Announcing intentions
- [ ] announce intentions in #status
  TODO: figure out wording

## Starting
- [ ] post to #status again right before starting
- [ ] update status.darklang.com (via betterstatus)

## Turning things off
- [ ] turn off pagerduty or phones
- [ ] turn off bwdserver (via k8s)   
- [ ] leave apiserver up (though, requests will largely fail upon DB connection)
- [ ] turn off cronchecker (turn crons down to 0 in kubernetes)
- [ ] update some LaunchDarkly settings
  - [ ] record what the values are currently, down below in "turning things back on"
  - [ ] `QUEUE_DELAY_BETWEEN_PULLS` -> 5s
  - [ ] QUEUE_MAX_CONCURRNECY_EVENTS_PER_WORKER -> 0

## Operations
- [ ] throughout, record activity in #operations
- [ ] sit with Google while they shrink the DB from 10TB
- [ ] reduce availability
- [ ] reduce CPU (16 vCPUs -> 12 vCPUs)
- [ ] (don't upgrade DB)

## Turning things back on
- [ ] update some LaunchDarkly settings
  - [ ] `QUEUE_DELAY_BETWEEN_PULLS` -> 
  - [ ] QUEUE_MAX_CONCURRNECY_EVENTS_PER_WORKER` -> 

## Announce done
- [ ] announce in #status

---

Later: set up a fancy status thing we can push messages to (to show at the top of darklang.com etc)
